src/main/java/hpi/CopyDataToWorkspacePlugin.java

```diff
@@ -79,15 +85,24 @@ public Environment setUp(AbstractBuild build, final Launcher launcher,
 		log.finest("Recognize project workspace and folder");
 		FilePath projectWorkspace = build.getWorkspace();
 		
-		Hudson hudson = Hudson.getInstance();
-		FilePath hudsonRoot = hudson.getRootPath();
+        Jenkins jenkins = Jenkins.get();
+        FilePath jenkinsRoot = jenkins.getRootPath();
 		
-		FilePath copyFrom = new FilePath(hudsonRoot, folderPath);
+		// Restrict access to userContent directory only
+		FilePath userContentDir = new FilePath(jenkinsRoot, "userContent");
+		FilePath copyFrom = new FilePath(userContentDir, folderPath);
 		
-		log.finest("Copying data from " + copyFrom.toURI()
-				+ " to " + projectWorkspace.toURI());
-		copyFrom.copyRecursiveTo(projectWorkspace);
+		// Verify if the final path is within allowed directory
+		if (!copyFrom.getRemote().startsWith(userContentDir.getRemote())) {

```

Comment:
`userContentDir.getRemote()` doesn't return the path with a trailing `/`, so if you're comparing it with another path using `.startsWith(...)`, you should append `File.separator` to ensure you're matching the exact directory and not another one with a similar prefix.

For example:
`userContent/`
& 
`userContentBis/`

Otherwise, you should also check to ensure there are no symlinks that could allow escaping the expected directory.